### ISSUE TITLE
Consider renaming r.do

### ISSUE THREAD

**Issue by danielmewes** [rethinkdb/rethinkdb#3542]
As has become clear in https://github.com/rethinkdb/rethinkdb/issues/2768 , it is very difficult to realize for users without extensive ReQL experience that `r.do` must be used instead of `r.map` for applying a function to a single datum.

We could rename `r.do` or introduce an alias for it that makes its similarity to `r.map` clearer.

@AtnNn has suggested `map1` (or `map_one`?). This would be the first ReQL term using a number in its name like that, but I personally think that it's really intuitive.

Other suggestions where `apply`, `with`, `call` and `let`, but all of those names seemed to have some problems.

**Comment by mglukhovsky** [comment]
I'm a fan of `map_one` -- I think it establishes `map` and `map_one` as pairs for mapping across either multiple or single documents. 

If the docs for each command reference both `map` and `map_one`, it should be very easy for novice users to pick the right command intuitively (which isn't the case with `do`).

**Comment by deontologician** [comment]
`pass_to`, though this works better for the infix version than the prefix version

**Comment by coffeemug** [comment]
:+1: for `map_one`

**Comment by AtnNn** [comment]
Although I like the name `do`, we could consider calling it `bind` or `as`. In other languages, `q.do(lambda x: b)` would otherwise be represented as `x = q; b;`. This might also work in ReQL, but it would duplicate `q`. It's the old lisp trick of using lambda to implement let.

I personally like `map1` a lot better than `map_one`, just because it is shorter.

**Comment by marshall007** [comment]
While `map_one` may sounds intuitive for single-selections I feel like it just flips the problem around and makes it unclear that it can still be used on selections. What about `value`, `to_value`, or `value_of`?

**Comment by deontologician** [comment]
Yeah, `r.expr(["a","b","c"]).map_one(...)` looks a bit misleading maybe?

**Comment by mglukhovsky** [comment]
Note that there's an open question: rename `do`, or alias it?

**Comment by danielmewes** [comment]
I'm tending towards creating an alias, because of what @marshall007 said.
It's not great that we will have two names for the same thing, but in this case it seems to make sense.

@deontologician Do you think people might assume that it maps only the first element out of the array?

**Comment by neumino** [comment]
Having `r.table("foo").get(1).map(...)` throws "`map` expects a sequence, got an object. Use `do` if you want to bind on a single value" would be pretty good too (maybe enough?).

The type error message are often hard to decipher for people.

**Comment by danielmewes** [comment]
Yeah I agree that we should change the error message for single selections to make them mention the "right" term (this was also suggested by @coffeemug in #2768).

In addition to that I still think that having a `map_one` alias to `do` will make the function a lot easier to discover for many users.

**Comment by deontologician** [comment]
I feel like aliases are a slippery slope (at least an alias we intend to keep permanently). Mayeb we should just sprinkle `.do` usage through examples more prominently so users are sure to run across it?

**Comment by mlucy** [comment]
I'm a little hesitant about `map1`/`map_one`.  When you start out knowing what it's supposed to do, it seems like an obvious name, but if you start out knowing the name I'm not sure it's obvious what it's supposed to do.  I could imagine a language where `map1` maps over one array and `map2` maps over two arrays, or where `map1` maps one function, etc. etc.

I'm also a little bit hesitant about aliasing, for all the obvious reasons, but I'd prefer aliasing to replacing `do` wholesale.

I think we should try to fix this with documentation and error messages.  Improving `map`'s error message and using `do` more frequently in the documentation might be enough.  If we do as much as we can along that axis and people are still confused, I think aliasing would be the logical next step.

**Comment by VeXocide** [comment]
I'd prefer not to create an alias unless we'll be deprecating and removing `do` in a number of releases, and agree we should improve the documentation first.

**Comment by danielmewes** [comment]
Ok, I'm open to the plan of improving error messages and documentation first and seeing how that works out.

**Comment by deontologician** [comment]
Just wanted to add a tidbit I ran across today. `.merge` will work on single selections unlike map. Seems like a strange inconsistency

**Comment by mlucy** [comment]
@deontologician -- that's actually the opposite case: `merge` in its basic form is defined on objects (but not on arrays), so when someone calls it on a sequence (whether that be a stream or an array) we can choose to map the merge over that sequence without introducing an ambiguity.  `do`, however, is defined on both objects *and* arrays in its basic form, so we can't automatically map it without introducing an ambiguity.
